File: src/com/anathema_roguelike/main/State.java
/*******************************************************************************
 * Copyright (C) 2017 Carl Minden
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/
package com.anathema_roguelike.main;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.PriorityQueue;

import com.anathema_roguelike.actors.Actor;
import com.anathema_roguelike.actors.TimeElapsedEvent;
import com.anathema_roguelike.entities.characters.Character;
import com.anathema_roguelike.entities.characters.inventory.PrimaryWeapon;
import com.anathema_roguelike.entities.characters.player.Player;
import com.anathema_roguelike.environment.Environment;
import com.anathema_roguelike.environment.generation.CaveDungeonGenerator;
import com.anathema_roguelike.environment.generation.DungeonGenerator;
import com.anathema_roguelike.main.display.Renderable;
import com.anathema_roguelike.main.ui.charactercreation.CharacterCreationUI;
import com.anathema_roguelike.main.utilities.position.Direction;
import com.anathema_roguelike.main.utilities.position.Point;
import com.anathema_roguelike.stats.characterstats.secondarystats.AttackSpeed;
import com.anathema_roguelike.stats.itemstats.WeaponSpeed;

public class State implements Renderable {
	
	private double elapsedTime = 0;
	double currentSegmentTime = 1;
	private Player player;
	private volatile ArrayList<Environment> dungeonLevels;
//	private DungeonGenerator dungeonLevelFactory = new DefaultDungeonLevelGenerator();
	private DungeonGenerator dungeonLevelFactory = new CaveDungeonGenerator();
//	private DungeonGenerator dungeonLevelFactory = new BigRoomDungeonGenerator();
	
	private PriorityQueue<Actor> actors = new PriorityQueue<Actor>(Collections.reverseOrder(new Comparator<Actor>() {

		@Override
		public int compare(Actor o1, Actor o2) {
			return Double.compare(o1.getEnergyLevel(), o2.getEnergyLevel());
		}
		
	}));
	
	public State() {
		
	}
	
	public void computeNextState() {
		
		Actor currentActor = actors.remove();
		
		while(!(currentActor instanceof Player)) {
			
			if(currentActor.getDuration().isExpired()) {
				currentActor = actors.remove();
				
				continue;
			}
			
			handleActor(currentActor);
			
			actors.add(currentActor);
			
			currentActor = actors.remove();
		}
		
		//Player CharacterAction
		handleActor(currentActor);
		actors.add(currentActor);
		
	}
	
	private void handleActor(Actor actor) {
		double startTime = actor.getEnergyLevel();
		
		if(actor.getEnergyLevel() <= 0) {
			actors.forEach(a -> a.energize());
		} else {
			actor.act();

			double timePassed = startTime - actors.peek().getEnergyLevel();
			
			elapsedTime += timePassed;
			currentSegmentTime -= timePassed;
			
			Game.getInstance().getEventBus().post(new TimeElapsedEvent(elapsedTime));
		}
	}
	
	public void render() {
		getCurrentEnvironment().render();
	}
	
	public Player getPlayer() {
		return player;
	}
	
	public Environment getCurrentEnvironment() {
		return player.getEnvironment();
	}

	public Environment getEnvironment(int z) {
		if(z < 0 || z >= Config.DUNGEON_DEPTH) {
			return null;
		}

		//TODO: WTF is this? is it to wait for another thread or something? seems like we could do better with a concurrency object
		while(dungeonLevels.size() <= z) {}
		
		return dungeonLevels.get(z);
	}
	
	public void generateDungeonLevels() {
		for(int i = 0; i < Config.DUNGEON_DEPTH; i++) {
			
			Environment newLevel = dungeonLevelFactory.createLevel(i);
			
			dungeonLevels.add(newLevel);
			
			for(Character character : newLevel.getEntities(Character.class)) {
				character.computeVisibility();
			}
		}
	}
	
	public double getElapsedTime() {
		return elapsedTime;
	}
	
	public void init() {
		dungeonLevels = new ArrayList<Environment>();
		generateDungeonLevels();
		
		player = CharacterCreationUI.createCharacter();
		
		Point upstairs = getEnvironment(0).getStairs(Direction.UP).getPosition();
		
		dungeonLevels.get(0).addEntity(player, upstairs);
	}
	
	public void registerActor(Actor actor) {
		actors.add(actor);
	}
	
	public void removeActor(Actor actor) {
		actors.remove(actor);
	}
}

File: src/com/anathema_roguelike/main/State.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package main

import java.util.{Collections, Comparator, PriorityQueue}

import com.anathema_roguelike.actors.{Actor, TimeElapsedEvent}
import com.anathema_roguelike.entities.characters.player.Player
import com.anathema_roguelike.environment.{Environment, Location}
import com.anathema_roguelike.environment.generation.{CaveDungeonGenerator, DungeonGenerator}
import com.anathema_roguelike.main.display.Renderable
import com.anathema_roguelike.main.ui.charactercreation.CharacterCreationUI
import com.anathema_roguelike.main.utilities.position.{Direction, Point}
import com.anathema_roguelike.entities.characters.Character

import scala.collection.mutable.ListBuffer

class State() extends Renderable {

  private var elapsedTime: Double = 0
  private var currentSegmentTime: Double = 1

  //	private DungeonGenerator dungeonLevelFactory = new DefaultDungeonLevelGenerator();
  private val dungeonLevelFactory: DungeonGenerator = new CaveDungeonGenerator
  //	private DungeonGenerator dungeonLevelFactory = new BigRoomDungeonGenerator();

  //TODO: Probably should move this to environment? doesnt make sense to simulate every level in the game (if we actually make more than one at a time)
  private val actors: PriorityQueue[Actor] = new PriorityQueue[Actor](Collections.reverseOrder(new Comparator[Actor]() {
    override def compare(o1: Actor, o2: Actor): Int = o1.getEnergyLevel.compareTo(o2.getEnergyLevel)
  }))

  private lazy val dungeonLevels = generateDungeonLevels()

  private lazy val upstairs: Location = getEnvironment(0).getStairs(Direction.UP).getLocation

  private lazy val player: Player = new Player(upstairs)

  def computeNextState(): Unit = {
    var currentActor: Actor = actors.remove

    while (!currentActor.isInstanceOf[Player]) {
      if (currentActor.getDuration.isExpired) {
        currentActor = actors.remove
      } else {
        handleActor(currentActor)
        actors.add(currentActor)
        currentActor = actors.remove
      }
    }

    handleActor(currentActor)
    actors.add(currentActor)
  }

  private def handleActor(actor: Actor): Unit = {
    val startTime: Double = actor.getEnergyLevel

    if (actor.getEnergyLevel <= 0) {
      actors.forEach((a: Actor) => a.energize())
    } else {
      actor.act()
      val timePassed: Double = startTime - actors.peek.getEnergyLevel

      elapsedTime += timePassed
      currentSegmentTime -= timePassed
      Game.getInstance.getEventBus.post(TimeElapsedEvent(elapsedTime))
    }
  }

  override def render(): Unit = {
    getCurrentEnvironment.render()
  }

  def getPlayer: Player = player

  def getCurrentEnvironment: Environment = player.getEnvironment

  def getEnvironment(z: Int): Environment = {
    if (z < 0 || z >= Config.DUNGEON_DEPTH) return null
    //TODO: WTF is this? is it to wait for another thread or something? seems like we could do better with a concurrency object
    while (dungeonLevels.size <= z) { }

    dungeonLevels(z)
  }

  def generateDungeonLevels(): Vector[Environment] = {

    (0 until Config.DUNGEON_DEPTH).toVector.map(i => {

      val newLevel: Environment = dungeonLevelFactory.createLevel(i)

      newLevel.getEntities[Character].foreach(c => c.computeVisibility())

      newLevel
    })
  }

  def getElapsedTime: Double = elapsedTime

  def registerActor(actor: Actor): Unit = {
    actors.add(actor)
  }

  def removeActor(actor: Actor): Unit = {
    actors.remove(actor)
  }
}
